src/kaleidoscope/plugin/FocusSerial.h

```diff
@@ -45,14 +45,14 @@ class FocusSerial : public kaleidoscope::Plugin {
     Kaleidoscope.serialPort().print(SEPARATOR);
   }
   template <typename Var, typename... Vars>
-  void send(Var v, const Vars&... vars) {
+  void send(Var v, Vars... vars) {
```

Comment:
This is an example of something small that should probably have just been its own PR.

Reply:
Right. But it's necessary for the virtual build only (causes problems only on x86 or with recent compilers - not sure).

Do you want me to factor it out?

Reply:
It's not a big deal and I don't need to throw the extra work on you for this one.

Ideally, PRs should be -small- and easy to review. Big architectural changes that include new features, changed APIs, and updates to every part of the code base to adapt to those new APIs are kind of a worst case for changes. Generally, anything that can be done to factor out changes that support future updates into their own PRs, even if that infrastructure isn't used yet, is a win. This isn't a hard and fast rule, but if functionality can be broken into logical stages, I'd rather review 25 20 line PRs than a 500 line PR. 

Obviously, for example, it doesn't make sense to break apart a PR that updates the codebase to change the calling convention from Key.flags to Key.getFlags(), but if the .raw stuff weren't a breaking change, it wouldn't be obvious to me that we shouldn't have broken apart the change to add the new API into its own PR and then the change to the usages into a second PR.

Reply:
> Generally, anything that can be done to factor out changes that support future updates into their own PRs, even if that infrastructure isn't used yet, is a win. This isn't a hard and fast rule, but if functionality can be broken into logical stages, I'd rather review 25 20 line PRs than a 500 line PR.

Thanks for clarifying this. I will try to adhere to the proposed workflow with separate PRs wherever possible.

But sometimes things are more complicated when working on larger changes that even require changes to multiple repos.

I understand you point from a reviewer's perspective. But from my personal contributor's perspective I find it more motivating, easier and efficient being allowed to submit larger PRs at least for greater changes. Also, whenever I submit something, I am prepared for it to be refused - which is also part of the game. Once that happens, I do not want to loose too much time that I already have invested on git-fiddling. I am a very fast coder but **very** slow when working with RCS. That's why I often code something, make it work and then submit it to obtain feedback as early as possible, sometimes even before taking too much time for cleaning up the git mess. I found that often to be more efficient because reviewers comments can require also a lot of rebasing work.

My typicall workflow when working on new features and larger changes (new plugin API, transient LED modes, host keymaps, virtual builds, ...) is to first do a lot of coding. I imagine how something works and then I just do it. Often in one go. Mostly with more or less frequent commits that only have backup character rather than being meant for submission. Sometimes ideas and work are evolving on the way. I am not the planing type but rather the lets-take-a-chissel-and-see-what's-inside-the-rock type. With such a proceeding it is extremely tedious and inefficient to plan ahead what to put into which commit. That's why I do not take too much care with commit messages during the creative stage of development. Finally, once work is finished and the implementation is solid, I clean up the mess and try to break down everything into independent commits. That can take many hours of rebasing and cherry picking - which is ok and part of the game. Nothing beats unbranched RCS trees.

This approach leads to PRs like the one I submitted here. It has a well defined aim - enabling virtual builds for all keyboards - and contains a lot of commits, each of which I treat as an independent change and try my very best to describe it as such.

If I would break this PR apart into pieces, it would possibly generate quite some PR interdependencies. Some of these PRs would need to be submitted sequentially. A dependent PR can only be submitted once all of its dependencies have been merged. Otherwise dependent PRs would need to contain commits of other dependee PRs. Submitting PRs sequencially requires extra bookkeeping and constantly keeping track of github merges.

As it is fairly normal for PRs to wait an undefined time in the review pipeline, chances are that I am not available for a while once a dependee PR has been merged. This stalls development until I find time to submit the next dependent PR that is possibly again a dependee of another PR...

Sometimes I am totally busy for a while. In other times I can spend quite a lot of spare time on OSS. In the latter case it feels good to do work, document and submit it as a whole.

@obra, I hope that explains a bit why and how I do what I do and I am open for suggestions about how I could improve my workflow to make both our lives easier?